Libraries/LibJS/Runtime/GeneratorFunction.cpp

```diff
@@ -0,0 +1,114 @@
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <AK/Function.h>
+#include <LibJS/AST.h>
+#include <LibJS/Interpreter.h>
+#include <LibJS/Runtime/Error.h>
+#include <LibJS/Runtime/GeneratorFunction.h>
+#include <LibJS/Runtime/GlobalObject.h>
+#include <LibJS/Runtime/Iterator.h>
+#include <LibJS/Runtime/Value.h>
+
+namespace JS {
+
+GeneratorFunction* GeneratorFunction::create(GlobalObject& global_object, const FlyString& name, const Statement& body, Vector<FlyString> parameters, LexicalEnvironment* parent_environment, LexicalEnvironment* own_environment)
+{
+    return global_object.heap().allocate<GeneratorFunction>(name, body, move(parameters), parent_environment, *global_object.function_prototype(), own_environment);
+}
+
+GeneratorFunction::GeneratorFunction(const FlyString& name, const Statement& body, Vector<FlyString> parameters, LexicalEnvironment* parent_environment, Object& prototype, LexicalEnvironment* own_environment)
+    : ScriptFunction(name, move(body), move(parameters), parent_environment, prototype)
+    , m_own_environment(own_environment)
+{
+}
+
+GeneratorFunction::~GeneratorFunction()
+{
+}
+
+void GeneratorFunction::visit_children(Visitor& visitor)
+{
+    ScriptFunction::visit_children(visitor);
+    visitor.visit(m_own_environment);
+}
+
+LexicalEnvironment* GeneratorFunction::create_environment()
+{
+    if (m_own_environment)
+        return m_own_environment;
+
+    return m_own_environment = ScriptFunction::create_environment();
+}
+
+Value GeneratorFunction::call(Interpreter& interpreter)
+{
+    auto& argument_values = interpreter.call_frame().arguments;
+    ArgumentVector arguments;
+    for (size_t i = 0; i < m_parameters.size(); ++i) {
+        auto name = parameters()[i];
+        auto value = js_undefined();
+        if (i < argument_values.size())
+            value = argument_values[i];
+        arguments.append({ name, value });
+        interpreter.current_environment()->set(name, { value, DeclarationKind::Var });
+    }
+
+    auto new_body = adopt(*new BlockStatement);
+    auto* block = static_cast<ScopeNode*>(new_body);
+
+    auto& my_block = static_cast<const ScopeNode&>(body());
+
+    for (auto& node : my_block.children()) {
+        block->append(node);
+    }
+    auto iterator = heap().allocate<Iterator>(heap().allocate<GeneratorFunction>(name(), *new_body, parameters(), m_parent_environment, *prototype(), create_environment()), [&interpreter, argument_values, arguments](Object* generator_object, const Vector<Value>& next_arguments) -> Iterator::IteratorResult {
+        auto* generator = static_cast<GeneratorFunction*>(generator_object);
```

Comment:
How do we know that `generator_object` is a `GeneratorFunction*`.
If it's always the case, can we communicate that through types instead of just blindly casting?

Reply:
In *this* specific callback, it's always `GeneratorFunction*`, but other users of Iterator may use any other Object.
I'm open to suggestions on how to do this more cleanly.